tests/dsl/dtd/dtd_test_ce.c

```diff
@@ -0,0 +1,683 @@
+int
+callback_tag_2(parsec_comm_engine_t *ce,
+               parsec_ce_tag_t tag,
+               void *msg,
+               size_t msg_size,
+               int src,
+               void *cb_data)
+{
+    (void) ce; (void) cb_data;
+    printf("[%d] In callback for tag %llu, message sent from %d size: %ld message: ", my_rank, tag, src, msg_size);
+
+    int i, total = msg_size/sizeof(int);
+
+    int *buffer = (int *)msg;
+    printf("[");
+    for(i = 0; i < total; i++) {
+        if(i == total - 1) {
+            printf("%d", buffer[i]);
+        } else {
+            printf("%d,", buffer[i]);
+        }
+    }
+    printf("]\n");
+    counter++;
+
+    return 1;
+}
+
+/* Tag 1 for int message */
+int
+callback_tag_3(parsec_comm_engine_t *ce,
+               parsec_ce_tag_t tag,
+               void *msg,
+               size_t msg_size,
+               int src,
+               void *cb_data)
+{
+    (void) ce; (void) cb_data;
+    printf("[%d] In callback for tag %llu, message sent from %d size: %ld message: ", my_rank, tag, src, msg_size);
+
+    int i, total = msg_size/sizeof(float);
+
+    float *buffer = (float *)msg;
+    printf("[");
+    for(i = 0; i < total; i++) {
+        if(i == total - 1) {
+            printf("%f", buffer[i]);
+        } else {
+            printf("%f,", buffer[i]);
+        }
+    }
+    printf("]\n");
+    counter++;
+
+    return 1;
+}
+
+typedef struct handshake_info_s {
+    int       buf_size;
+    uintptr_t cb_fn;
+} handshake_info_t;
+
+
+/* Active Message for GET notification.
+ * This function will be triggered in rank 1 after rank 0
+ * sends an active message to 1 informing about a GET that
+ * 1 needs to complete.
+ */
+int
+notify_about_get(parsec_comm_engine_t *ce,
+                 parsec_ce_tag_t tag,
+                 void *msg,
+                 size_t msg_size,
+                 int src,
+                 void *cb_data)
+{
+    (void) tag; (void) cb_data; (void) msg_size;
+    assert(my_rank == 1);
+
+    handshake_info_t *GET_activation_message = (handshake_info_t *) msg;
+
+    /* We have the remote memory_handle.
+     * Let's allocate the local memory_handle
+     * and let's start the GET.
+     */
+    parsec_ce_mem_reg_handle_t rank_1_memory_handle;
+    size_t rank_1_memory_handle_size;
+    /* GET operation will store the actual data in the following buffer */
+    int *receive_buf = malloc(sizeof(int) * GET_activation_message->buf_size);
+
+    if(ce->capabilites.supports_noncontiguous_datatype) {
+        parsec_datatype_t *datatype = malloc(sizeof(parsec_datatype_t));
+        parsec_type_create_contiguous(GET_activation_message->buf_size, parsec_datatype_int_t, datatype);
+        ce->mem_register(receive_buf, PARSEC_MEM_TYPE_NONCONTIGUOUS,
+                         1, *datatype,
+                         -1,
+                         &rank_1_memory_handle, &rank_1_memory_handle_size);
+    } else {
+         ce->mem_register(receive_buf, PARSEC_MEM_TYPE_CONTIGUOUS,
```

Comment:
Reading the code in `parsec_mpi_funnelled.c`, I don't see any places where the datatype can be -1. Some compilers complain that we cast an int (-1) into a pointer, which is right (to complain). Did I miss where `(void*)-1` is a special datatype constant? Should we define it (this issue occurs 4 times in this file, I will add a comment for each.

Reply:
`PARSEC_DATATYPE_NULL` should be used, not `(void*)(-1)`. This was supposed to separate out cases where the comm engine backend supports non-contiguous datatypes (MPI) and that don't (LCI), but splitting up the logic in `mem_register` isn't really necessary—it's important that the comm engine exposes its capabilities, but computing the size of data as size of datatype * count can be done inside of `mem_register`.

Reply:
Oh I see now you've already fixed that. :)